tools/mtouch/BuildTasks.mtouch.cs

```diff
@@ -191,36 +191,57 @@ public class AOTTask : ProcessTask
 			}
 		}
 
+		IEnumerable<Assembly> InputAssemblies {
+			get {
+				var accum = new List<Assembly>();
+				accum.Add (Assembly);
+
+				if (Assembly.IsDedupDummy) {
+					// The dedup assembly AOT emits a file with
+					// the name of DedupAssembly so the output
+					// is the same, but the input is now every assembly
+					accum.Union (Assembly.Target.Assemblies);
+				}
+
+				return accum;
+			}
+		}
+
 		public override IEnumerable<string> Inputs {
 			get {
-				yield return Assembly.FullPath;
+				return InputAssemblies.Select ((v) => v.FullPath);
 			}
 		}
 
 		public override IEnumerable<string> FileDependencies {
 			get {
 				if (inputs == null) {
 					inputs = new List<string> ();
-					if (Assembly.HasDependencyMap)
-						inputs.AddRange (Assembly.DependencyMap);
-					inputs.Add (AssemblyName);
-					inputs.Add (Driver.GetAotCompiler (Assembly.App, Assembly.Target.Is64Build));
-					var mdb = Assembly.FullPath + ".mdb";
-					if (File.Exists (mdb))
-						inputs.Add (mdb);
-					var pdb = Path.ChangeExtension (Assembly.FullPath, ".pdb");
-					if (File.Exists (pdb))
-						inputs.Add (pdb);
-					var config = Assembly.FullPath + ".config";
-					if (File.Exists (config))
-						inputs.Add (config);
+					foreach (var ass_dep in InputAssemblies) {
+						if (ass_dep.HasDependencyMap)
+							inputs.AddRange (ass_dep.DependencyMap);
+						inputs.Add (ass_dep.FullPath);
+						inputs.Add (Driver.GetAotCompiler (ass_dep.App, ass_dep.Target.Is64Build));
+						var mdb = ass_dep.FullPath + ".mdb";
+						if (File.Exists (mdb))
+							inputs.Add (mdb);
+						var pdb = Path.ChangeExtension (ass_dep.FullPath, ".pdb");
+						if (File.Exists (pdb))
+							inputs.Add (pdb);
+						var config = ass_dep.FullPath + ".config";
+						if (File.Exists (config))
+							inputs.Add (config);
+					}
 				}
 				return inputs;
 			}
 		}
 
 		public override bool IsUptodate {
 			get {
+				if (Assembly.IsDedupDummy)
+					return false;
```

Comment:
Strictly speaking, the contents of the dedup dummy amodule will end up having to be recompiled every time that any other assembly changes. I could make this case depend on the IsUptodate property of every other assembly provided.

Reply:
Can I force this to always be rebuilt if we rebuild, but not trigger a rebuild?

Reply:
@rolfbjarne any suggestion ?

Reply:
The best way to do this depends on whether the dedup assembly is needed at runtime or not, which I already asked here: https://github.com/xamarin/xamarin-macios/pull/3783#discussion_r175665845

Reply:
It’s very necessary, as loading it at runtime loads all of the deduplicated inflated methods and wrappers. It is a container for all AOT code that does not have a 1-1 mapping with a managed method.

Reply:
I think @rolfbjarne meant the metadata, not the AOT code from that assembly (which ends up in the native executable). My guess is that it's needed, along with the .aotdata file, but my understanding of it could be wrong :)

Reply:
Ah, I guess the aotdata is needed, because the assembly itself contains nothing interesting, and the native code ends up in the executable.

Reply:
This should not be needed as long as `FileDependencies` list all the assemblies when AOT-compiling the dedup assembly (which it looks like it does).